### PR TITLE
test: Fix TestMachinesCreate.testCreate{PXE,AndVerifyQemuConf}

### PR DESCRIPTION
**Pull request by martinpitt**
Destroying pxe-guest does not immediatel stop virt-install. Instead,
this interferes with our redefinition and restarting of the domain, so
that sometimes either failed. This resulted in the domain sometimes not
actually having the /tmp/serial.txt console, or already running when
the test tried to start it.

----

This fails [very](https://logs.cockpit-project.org/logs/pull-631-20220322-175241-f9703f15-arch/log.html#65-1) [often](https://logs.cockpit-project.org/logs/pull-630-20220322-131300-6b7ecc3e-fedora-36/log.html#13) like this:
```
sed: can't read /tmp/serial.txt: No such file or directory
Traceback (most recent call last):
  File "/usr/lib64/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/work/bots/make-checkout-workdir/test/check-machines-create", line 321, in testCreatePXE
    wait(lambda: self.machine.execute(r"sed 's,\x1B\[[0-9;]*[a-zA-Z],,g' /tmp/serial.txt | grep 'Rebooting in 60'"), delay=3)
  File "/work/bots/make-checkout-workdir/test/common/testlib.py", line 2131, in wait
    val = func()
  File "/work/bots/make-checkout-workdir/test/check-machines-create", line 321, in <lambda>
    wait(lambda: self.machine.execute(r"sed 's,\x1B\[[0-9;]*[a-zA-Z],,g' /tmp/serial.txt | grep 'Rebooting in 60'"), delay=3)
  File "/work/bots/make-checkout-workdir/bots/machine/machine_core/ssh_connection.py", line 368, in execute
    raise subprocess.CalledProcessError(proc.returncode, command, output=output)
subprocess.CalledProcessError: Command 'sed 's,\x1B\[[0-9;]*[a-zA-Z],,g' /tmp/serial.txt | grep 'Rebooting in 60'' returned non-zero exit status 1.
```

i.e. /tmp/serial.txt never gets created. There seems to be a race condition when defining the VM, as the VNC console in the [screenshot](https://logs.cockpit-project.org/logs/pull-631-20220322-175241-f9703f15-arch/TestMachinesCreate-testCreatePXE-arch-127.0.0.2-2401-FAIL.png)  shows the successful PXE boot just fine.

I can't reproduce this locally, so debugging on the infra.

 - Builds on top of PR #631

### CODE REVIEW

**Comment by martinpitt** [comment]
[one hit](https://logs.cockpit-project.org/logs/pull-633-20220323-060023-53751841-fedora-35-firefox/log.html#12), with the [relevant log](https://github.com/cockpit-project/cockpit-machines/files/8330257/log.txt). This confirms that pxe-guest does *not* have the serial console defined. It also still has all the default `<console>` objects, so the XML manipulation did not work, or the re-definition of the domain silently failed. I'll add some more debugging to show the generated /tmp/domain.xml, to see which one it is.

**Comment by martinpitt** [comment]
Interesting -- now there are several failures [like this](https://logs.cockpit-project.org/logs/pull-633-20220323-064845-409ea6b4-rhel-9-0/log.html#17), where the dumpxml from the regenerated domain is correct, but trying to start it fails:
```
+ virsh start pxe-guest
error: Domain is already active

Traceback (most recent call last):
  File "/work/bots/make-checkout-workdir/test/check-machines-create", line 329, in testCreatePXE
    self.machine.execute("virsh start pxe-guest")
```
I haven't seen this before, so I find it relatively unlikely that it happened before (as the original failure happens later on). So somehow undefining the domain triggered a (re)start of the domain. I [added a commit](https://github.com/cockpit-project/cockpit-machines/pull/633/commits/1d2ffc6201634f605de232ff61680bba88a0db70) which waits for the `destroy` to actually happen -- it might be async?

**Comment by martinpitt** [comment]
Bazinga! [all passed](https://github.com/cockpit-project/cockpit-machines/commit/9205bb52db3bb3318f2be13ca07d74a788137887) on that commit. I'll now push a cleaned-up version of this.

**Comment by martinpitt** [comment]
All good in the previous run, except two instances of [testCreateAndVerifyQemuConf](https://logs.cockpit-project.org/logs/pull-633-20220323-095551-43cedf87-centos-8-stream/log.html#64) [failure](https://logs.cockpit-project.org/logs/pull-633-20220323-095551-43cedf87-debian-testing/log.html#64-1). That's unrelated -- I'll look at this, but not as part of this PR.

**Comment by martinpitt** [comment]
That [testCreateAndVerifyQemuConf](https://logs.cockpit-project.org/logs/pull-633-20220323-103314-d765fc14-centos-8-stream/log.html#59) test keeps failing, and the reason is so obvious that I'll just fix it here right away after all.

File: test/check-machines-create [hunk 1]
```diff
@@ -313,7 +313,8 @@ class TestMachinesCreate(VirtualMachinesCase):
         # Redefine the domain with the new XML
         xmlstr = ET.tostring(root, encoding='unicode', method='xml')
 
-        self.machine.execute(f"echo '{xmlstr}' > /tmp/domain.xml && virsh define --file /tmp/domain.xml")
+        self.write_file("/tmp/domain.xml", xmlstr)
+        self.machine.execute("virsh define --file /tmp/domain.xml")
```
Comment (skobyda) [inline test/check-machines-create:320]: So the problem here, if I understand it correctly was that virsh define was reading the old file domain.xml?

File: test/check-machines-create [hunk 2]
```diff
@@ -275,9 +275,12 @@ class TestMachinesCreate(VirtualMachinesCase):
         # We don't want to use start_vm == False because if we get a separate install phase
         # virt-install will overwrite our changes.
 
-        # Wait for virt-install to define the VM and then stop it
-        wait(lambda: "pxe-guest" in self.machine.execute("virsh list --persistent"), delay=3)
+        # Wait for virt-install to define the VM, stop it, and wait for virt-install to be done
+        wait(lambda: "pxe-guest" in self.machine.execute("virsh list --name --persistent"), delay=3)
+        wait(lambda: "pxe-guest" in self.machine.execute("virsh list --name"), delay=3)
         self.machine.execute("virsh destroy pxe-guest")
+        self.machine.execute("while pgrep virt-install; do sleep 1; done")
```
Comment (skobyda) [inline test/check-machines-create:282]: Is this correct. Here we are waiting until virt-install to finish and process to cease existing, right?
But according to manpages, pgrep return 0 when process exists, and 1 when it doesn't exist, so the while condition should be negated, no?
```
EXIT STATUS
       0      One or more processes matched the criteria. For pkill and pidwait, one or more processes must also have been successfully signalled or waited for.
       1      No processes matched or none of them could be signalled.
```

File: test/check-machines-create (hunk 1 above)
Comment (martinpitt) [reply test/check-machines-create:320]: No, /tmp/domain.xml has always been okay. But the running virt-install in the background and this `virsh define` collided, and sometimes virt-install won -- hence this `virsh define` was a no-op.

File: test/check-machines-create (hunk 2 above)
Comment (martinpitt) [reply test/check-machines-create:282]: Right, the point of that is to wait while virt-install is running -- i.e. wait until it is *not* running any more. So this works as intended.

Comment (martinpitt) [reply test/check-machines-create:282]: No, it's already running and it works until it stops running. Precisely, while it is running (pgrep finds a process) it sleeps.

**Pull request merged**